milli/src/update/index_documents/helpers/grenad_helpers.rs

```diff
@@ -0,0 +1,276 @@
+    })?;
+    merger.write_into(&mut writer)?;
+    let reader = writer_into_reader(writer)?;
+    Ok(reader)
+}
+
+#[derive(Debug, Clone, Copy)]
+pub struct GrenadParameters {
+    pub chunk_compression_type: CompressionType,
+    pub chunk_compression_level: Option<u32>,
+    pub max_memory: Option<usize>,
+    pub max_nb_chunks: Option<usize>,
+}
+
+impl Default for GrenadParameters {
+    fn default() -> Self {
+        Self {
+            chunk_compression_type: CompressionType::None,
+            chunk_compression_level: None,
+            max_memory: None,
+            max_nb_chunks: None,
+        }
+    }
+}
+
+impl GrenadParameters {
+    pub fn max_memory_by_thread(&self) -> Option<usize> {
+        self.max_memory.map(|max_memory| max_memory / rayon::current_num_threads())
+    }
+}
+
+/// Returns an iterator that outputs grenad readers of obkv documents
+/// with a maximum size of approximately `documents_chunks_size`.
+///
+/// The grenad obkv entries are composed of an incremental document id big-endian
+/// encoded as the key and an obkv object with an `u8` for the field as the key
+/// and a simple UTF-8 encoded string as the value.
+pub fn grenad_obkv_into_chunks<R: io::Read>(
+    mut reader: grenad::Reader<R>,
+    indexer: GrenadParameters,
+    log_frequency: Option<usize>,
+    documents_chunk_size: Byte,
+) -> Result<impl Iterator<Item = Result<grenad::Reader<File>>>> {
+    let mut document_count = 0;
+    let mut continue_reading = true;
+
+    let indexer_clone = indexer.clone();
+    let mut transposer = move || {
+        if !continue_reading {
+            return Ok(None);
+        }
+
+        let mut current_chunk_size = 0u64;
+        let mut obkv_documents = tempfile::tempfile().and_then(|file| {
+            create_writer(
+                indexer_clone.chunk_compression_type,
+                indexer_clone.chunk_compression_level,
+                file,
+            )
+        })?;
+
+        while let Some((document_id, obkv)) = reader.next()? {
+            obkv_documents.insert(document_id, obkv)?;
+            current_chunk_size += document_id.len() as u64 + obkv.len() as u64;
+
+            document_count += 1;
+            if log_frequency.map_or(false, |log_frequency| document_count % log_frequency == 0) {
+                debug!("reached {} chunked documents", document_count);
+            }
+
+            if current_chunk_size >= documents_chunk_size.get_bytes() {
+                return writer_into_reader(obkv_documents).map(Some);
+            }
+        }
+
+        continue_reading = false;
+        writer_into_reader(obkv_documents).map(Some)
+    };
+
+    Ok(std::iter::from_fn(move || {
+        let result = transposer().transpose();
+        if result.as_ref().map_or(false, |r| r.is_ok()) {
+            debug!(
+                "A new chunk of approximately {} has been generated",
+                documents_chunk_size.get_appropriate_unit(true),
+            );
```

Comment:
Do we want to keep this log?

Reply:
I find it useful to know when in the indexing timeline we generate chunks. Until we have a tool like puffin to visualize it, I'd prefer to keep this kind of log.